src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -472,8 +441,37 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
                     goto ConcurrentOperation;
                 }
             }
+            else
+            {
+                uint hashCode = (uint)comparer.GetHashCode(key);
+                int i = GetBucket(hashCode);
+                Entry[] entries = _entries;
+                uint collisionCount = 0;
+                i--; // Value in _buckets is 1-based; subtract 1 from i. We do it here so it fuses with the following conditional.
+                do
+                {
+                    // Should be a while loop https://github.com/dotnet/runtime/issues/9422
+                    // Test in if to drop range check for following array access
+                    if ((uint)i >= (uint)entries.Length)
+                    {
+                        goto ReturnNotFound;
+                    }
+
+                    entry = ref entries[i];
+                    if (entry.hashCode == hashCode && comparer.Equals(entry.key, key))
+                    {
+                        goto ReturnFound;
+                    }
 
-            goto ReturnNotFound;
```

Comment:
Still want this goto or its going to fail with the wrong message for things not found?